pkg/chunked/storage_linux.go

```diff
@@ -1579,6 +1584,9 @@ func (c *chunkedDiffer) ApplyDiff(dest string, options *archive.TarOptions) (gra
 	if totalChunksSize > 0 {
 		logrus.Debugf("Missing %d bytes out of %d (%.2f %%)", missingPartsSize, totalChunksSize, float32(missingPartsSize*100.0)/float32(totalChunksSize))
 	}
+
+	output.UncompressedDigest = c.uncompressedDigest
```

Comment:
Actually, this seems to be a major problem.

We rely on `UncompressedDigest` for layer deduplication; if we accept this value just provided by the caller at face value, it seems to me that an attacker could create a Zstd:chunked layer with malicious contents, and _claim_ in the metadata that the `UncompressedDigest` is a value that matches a known widely-used base image. That could then cause the malicious contents to be used for that layer, affecting pulls of other (correctly-signed) containers.

For manifest digest matching (and therefore signature validation) rule is that we only use digests that we compute ourselves (i.e. the compressed-layer digests we always compute and match; we also compute uncompressed-layer digests ourselves, and record the compressed/uncompressed pairs _locally_ to allow safe layer reuse later).

Ugh… I suppose we would need to have `ApplyDiffFromStagingDirectory` actually recreate the uncompressed tarball and to compute the full digest? That seems awfully expensive.

Reply:
indeed.  Thanks for spotting it!

What do you think if we change the logic for `digestingReader` to not raise an error when the digest is computed but just use the correct digest instead of the one we store (that could be the compressed digest)?  Maybe we could limit this behavior only for the partial layers?

Reply:
I was thinking of something like the following for c/image:

```diff
diff --git a/copy/blob.go b/copy/blob.go
index f45b97f5..293ed4f1 100644
--- a/copy/blob.go
+++ b/copy/blob.go
@@ -129,7 +129,9 @@ func (ic *imageCopier) copyBlobFromStream(ctx context.Context, srcReader io.Read
        }
 
        if digestingReader.validationFailed { // Coverage: This should never happen.
-               return types.BlobInfo{}, fmt.Errorf("Internal error writing blob %s, digest verification failed but was ignored", srcInfo.Digest)
+               uploadedInfo.Digest = digestingReader.digester.Digest()
+               logrus.Warningf("Digest verification failed for blob %s, using computed %s", srcInfo.Digest, uploadedInfo.Digest)
+               return uploadedInfo, nil
        }
        if stream.info.Digest != "" && uploadedInfo.Digest != stream.info.Digest {
                return types.BlobInfo{}, fmt.Errorf("Internal error writing blob %s, blob with digest %s saved with digest %s", srcInfo.Digest, stream.info.Digest, uploadedInfo.Digest)
diff --git a/copy/digesting_reader.go b/copy/digesting_reader.go
index 901d1082..028c86fc 100644
--- a/copy/digesting_reader.go
+++ b/copy/digesting_reader.go
@@ -54,9 +54,7 @@ func (d *digestingReader) Read(p []byte) (int, error) {
                actualDigest := d.digester.Digest()
                if actualDigest != d.expectedDigest {
                        d.validationFailed = true
-                       return 0, fmt.Errorf("Digest did not match, expected %s, got %s", d.expectedDigest, actualDigest)
                }
-               d.validationSucceeded = true
        }
        return n, err
 }
```

so we won't have to pay the high price of computing the uncompressed digest at pull time (which would probably defeat all the advantages of a partial pull)

Reply:
I don’t think that code path is involved in a partial pull at all. What am I missing?!

(Also, `validationFailed` is when we _positively know_ the digest does not match, because we have read all of the blob; I don’t see why it would ever be OK to continue in that situation.)